keadm/cmd/keadm/app/cmd/upgrade.go

```diff
@@ -13,22 +13,23 @@ Specify whether to upgrade the cloud or the edge through three-level commands.
 If no three-level command, it upgrades edge components.`
 )
 
+// NewUpgradeCommand creates a upgrade command instance and returns it.
 func NewUpgradeCommand() *cobra.Command {
 	cmds := &cobra.Command{
 		Use:   "upgrade",
 		Short: "Upgrade components of the cloud or the edge",
 		Long:  upgradeLongDescription,
 	}
 
-	edgecmd := edge.NewEdgeUpgrade()
-
 	// Used for backward compatibility of the edgecore trigger the upgrade command
 	upgradeOptions := edge.NewUpgradeOptions()
+	cmds.RunE = func(cmd *cobra.Command, args []string) error {
+		return upgradeOptions.Upgrade()
+	}
 	edge.AddUpgradeFlags(cmds, upgradeOptions)
-	cmds.RunE = edgecmd.RunE
 
```

Comment:
Sorry, I'm a little confused about the modifications here. Why do we need to assign a separate value to cmd.RunE? It seems that the original method can also achieve the same effect.

Reply:
That's what I thought before, so I assigned `edgecmd.RunE` to `cmd.RunE`. If we look closely, we can see that the upgradeOptions is not passed to the function edgecmd.RunE() in any way, so where do the upgradeOptions come from, that's the scope of the closure. The upgradeOptions is assigned to the closure function RunE() after initialization in the NewEdgeUpgrade(), so the closure RunE use the upgradeOptions in the NewEdgeUpgrade()
```golang
func NewEdgeUpgrade() *cobra.Command {
    opts := NewUpgradeOptions()  // address: 0x01
    RunE: func(...) {
        opts.Upgrade() // opts address: 0x01
    }
}

func NewCloudUpgrade() *cobra.Command {
    edgeCmd := NewEdgeUpgrade()
    opts := newCloudUpgradeOptions() // address: 0x02
    RunE: func(...) {
        opts.Upgrade() // opts address: 0x02
    }
    RunE = edgeCmd.RunE = func(...){ opts.Upgrade() /* opts address: 0x01 */ }
}
```

Reply:
LGTM.
I understand the reason for your modification here. It is true that RunE in the original code uses the UpgradeOptions defined in NewEdgeUpgrade() instead of in NewUpgradeCommand(). This bug is really hard to find. 
I have another question, what will this bug perform when it is actually run? Did some subsequent module update the UpgradeOptions in NewUpgradeCommand(), but the update did not take effect because we use the Options of NewEdgeUpgrade() before?

Reply:
Yes, edge nodes that use the default upgrade command `keadm upgrade` will fail to upgrade, because the UpgradeOptions instantce is just a default value created by the function edge.NewUpgradeOptions()

Reply:
OK, thanks for your detailed answer!
/lgtm
cc @Shelley-BaoYue @fisherxu